compiler/rustc_mir_dataflow/src/move_paths/builder.rs

```diff
@@ -113,21 +113,21 @@ impl<'b, 'a, 'tcx> Gatherer<'b, 'a, 'tcx> {
         // from `*(u.f: &_)` isn't allowed.
         let mut union_path = None;
 
-        for (i, elem) in place.projection.iter().enumerate() {
-            let proj_base = &place.projection[..i];
+        for (place_ref, elem) in place.as_ref().iter_projections() {
             let body = self.builder.body;
             let tcx = self.builder.tcx;
-            let place_ty = Place::ty_from(place.local, proj_base, body, tcx).ty;
+            let place_ty = place_ref.ty(body, tcx).ty;
+
             match place_ty.kind() {
                 ty::Ref(..) | ty::RawPtr(..) => {
-                    let proj = &place.projection[..i + 1];
                     return Err(MoveError::cannot_move_out_of(
                         self.loc,
                         BorrowedContent {
                             target_place: Place {
                                 local: place.local,
-                                projection: tcx.mk_place_elems(proj),
-                            },
+                                projection: tcx.mk_place_elems(place_ref.projection),
+                            }
+                            .project_deeper(&[elem], tcx),
```

Comment:
Could you add a `project_deeper` method on `PlaceRef`, to avoid interning twice?

Reply:
Or should `BorrowedContent` take a `PlaceRef`?

Reply:
I think `PlaceRef` already had that. I'm not sure I'm really qualified to have a take on what `BorrowedContent` should take but even if it took a `PlaceRef` we'd need some way to produce the extended `PlaceRef`. We could also try building the iterator so that we have access to the subsequent projection in each iteration, but I'm not sure how worth it that is.